gym/envs/registration.py

```diff
@@ -78,558 +75,191 @@ def parse_env_id(id: str) -> Tuple[Optional[str], str, Optional[int]]:
-        ):
-            logger.warn(
-                f"The environment {path} is out of date. You should consider "
-                f"upgrading to version `v{latest_versioned_spec.version}` "
-                f"with the environment ID `{latest_versioned_spec.id}`."
-            )
-        elif latest_versioned_spec and version is None:
-            logger.warn(
-                f"Using the latest versioned environment `{latest_versioned_spec.id}` "
-                f"instead of the unversioned environment `{path}`"
-            )
-            path = latest_versioned_spec.id
+    # Process possible versioned environments
 
-        # Lookup our path
-        spec = self.spec(path)
-        # Construct the environment
-        return spec.make(**kwargs)
+    versioned_specs = [spec_ for spec_ in env_specs if spec_.version is not None]
 
-    def all(self):
-        return self.env_specs.values()
+    latest_spec = max(versioned_specs, key=lambda spec: spec.version, default=None)  # type: ignore
+    if latest_spec is not None and version > latest_spec.version:
+        version_list_msg = ", ".join(f"`v{spec_.version}`" for spec_ in env_specs)
+        message += f" It provides versioned environments: [ {version_list_msg} ]."
 
-    def spec(self, path: str) -> EnvSpec:
-        if ":" in path:
-            mod_name, _, id = path.partition(":")
-            try:
-                importlib.import_module(mod_name)
-            except ModuleNotFoundError:
-                raise error.Error(
-                    f"A module ({mod_name}) was specified for the environment but was not found, "
-                    "make sure the package is installed with `pip install` before calling `gym.make()`"
-                )
-        else:
-            id = path
+        raise error.VersionNotFound(message)
 
-        # We can go ahead and return the env_spec.
-        # The EnvSpecTree will take care of any exceptions.
-        return self.env_specs[id]
+    if latest_spec is not None and version < latest_spec.version:
+        raise error.DeprecatedEnv(
+            f"Environment version v{version} for `{get_env_id(ns, name, None)}` is deprecated. "
+            f"Please use `{latest_spec.id}` instead."
+        )
 
-    def register(self, id: str, **kwargs) -> None:
-        spec = EnvSpec(id, **kwargs)
 
-        if self._ns is not None:
-            if spec.namespace is not None:
-                logger.warn(
-                    f"Custom namespace `{spec.namespace}` is being overridden "
-                    f"by namespace `{self._ns}`. If you are developing a "
-                    "plugin you shouldn't specify a namespace in `register` "
-                    "calls. The namespace is specified through the "
-                    "entry point package metadata."
-                )
-            # Replace namespace
-            spec.namespace = self._ns
+def find_newest_version(ns: Optional[str], name: str) -> Optional[int]:
```

Comment:
docstring

Reply:
Very minor point, would `find_highest_version` be a more accurate name? Newest is possibly confusing as you could think the latest spec to be added

Reply:
Renamed